src/util/tmpl/fd_treap.c

```diff
@@ -776,6 +789,91 @@ TREAP_(idx_remove)( TREAP_(t) * treap,
   return treap;
 }
 
+/* Merges the (sub)treap rooted at the node with index a with the
+   (sub)treap rooted at the node with index b.  Returns the index of the
+   root node of the treap containing their merge.  The value of
+   TREAP_PARENT of the returned node is undefined. */
+TREAP_IDX_T
+TREAP_(private_merge)( TREAP_IDX_T  a,
+                       TREAP_IDX_T  b,
+                       TREAP_T    * pool ) {
+
+  if( TREAP_IDX_IS_NULL( a ) ) return b;
+  if( TREAP_IDX_IS_NULL( b ) ) return a;
+
+  ulong a_prio = (ulong)pool[ a ].TREAP_PRIO;
+  ulong b_prio = (ulong)pool[ b ].TREAP_PRIO;
+
+  /* By hypothesis, a's priority is the highest in subtreap a and b's
+     priority is the highest in subtreap b.  That means the one with the
+     higher priority will have the highest priority in the merged
+     subtreap, so it will be the new root. */
+  int new_root_is_a = (a_prio>b_prio) | ((a_prio==b_prio) & ((int)(a ^ b) & 1));
+  TREAP_IDX_T new_root = (TREAP_IDX_T)fd_ulong_if( new_root_is_a, a, b );
+  TREAP_IDX_T non_root = (TREAP_IDX_T)fd_ulong_if( new_root_is_a, b, a );
+
+  /* If the new root is a, then we want to split subtreap b on a's key.
+     Since the pool may be full, we can't necessarily acquire a
+     temporary node and insert it, so the easiest way to do this is to
+     make a temporary treap from the subtreap rooted at b and use the
+     normal insert method to insert a into it.  We need to be careful
+     that we don't lose any important links in the process, so we cache
+     them. */
+
+  TREAP_IDX_T root_l  = pool[ new_root ].TREAP_LEFT;
+  TREAP_IDX_T root_r  = pool[ new_root ].TREAP_RIGHT;
+  TREAP_IDX_T root_pr = pool[ new_root ].TREAP_PRIO;
+
+
+  pool[ non_root ].TREAP_PARENT = TREAP_IDX_NULL;
+  /* We know new_root's priority is >= all other priorities in subtreap
+     b, but it might lose some random ties, so we set it to the max.  We
+     need to make sure it ends up in the root of the subtreap. */
+  pool[ new_root ].TREAP_PRIO   = (TREAP_IDX_T)ULONG_MAX;
+
+  TREAP_(t) temp[1];
+  temp->root = non_root;
+
+  TREAP_(idx_insert)( temp, new_root, pool );
+
+  FD_TEST( temp->root == new_root );
+
+  /* Now we have two subtreaps that belong to the left of new_root and
+     two that belong to the right, so we merge them recursively and make
+     new_root the root. */
+
+  TREAP_IDX_T nonr_l  = pool[ new_root ].TREAP_LEFT;
+  TREAP_IDX_T nonr_r  = pool[ new_root ].TREAP_RIGHT;
+
+  TREAP_IDX_T new_l = TREAP_(private_merge)( root_l, nonr_l, pool );
+  TREAP_IDX_T new_r = TREAP_(private_merge)( root_r, nonr_r, pool );
```

Comment:
You are right that the worst case recursion depth is O(number of nodes):

> Consider the case in which `treap_a` and `treap_b` both only have right children, each node in `treap_a` is less than all nodes in `treap_b`, and the priority of each node in `treap_a` is higher than the priority of all nodes in `treap_b`.
`new_root` is always chosen from `treap_a`. When `treap_b` is split on the new root, the entire treap becomes a right child of the dummy node. 
The first recursive call (`new_l`) merges two empty treaps. For the second recursive call (`new_r`), `root_r` has one fewer node than `treap_a` started with, and `nonr_r` is the entirety of `treap_b`.

I was really hoping there was a non-recursive way to do this, similar to the way insert is done, but I couldn't find a way to get around needing O(height) space. Each stack frame is relatively small (just a couple of ulongs).

Let me think about this. I just realized you don't have to save the original priority, because you could increment it (safe given then half-open interval) and then decrement it after the split.

Reply:
Vaguely recall pondering merge support when whipping up the treap earlier.  IIRC, I was pretty sure a safe implementation was possible.  But it also looked tricky, I didn't have any immediate need for it and there wasn't always fast analogous algorithms for other tree-like search structures ( so wasn't obvious users would expect such either).  So I just kicked the can.

It almost surely is possible if we add extra auxiliary fields to the element to hold the recursion state.  If aux fields are needed, I would probably give compile time options for merge support so non-merging uses don't have spend memory / cache for stuff they don't need.

It is plausible that recursion state could be done by careful use of existing fields (like the prio_max +/- 1 trick you noted ... noticed this too but didn't connect it to the recursion implications and thought the splitting was easier to follow without the trick).